test/protocol/variables_test.rb

```diff
@@ -0,0 +1,85 @@
+# frozen_string_literal: true
+
+require_relative "../support/protocol_test_case"
+
+module DEBUGGER__
+  class DAPVariablesTest < ProtocolTestCase
+    PROGRAM = <<~RUBY
+      1| $a = 1
+      2| $b = 2
+      3| $c = 3
+    RUBY
+
+    def test_eval_evaluates_global_variables
+      run_protocol_scenario PROGRAM, cdp: false do
+        req_add_breakpoint 3
+        req_continue
+
+        globals = gather_variables(type: "globals")
+
+        # User defined globals
+        assert_includes(globals, { name: "$a", value: "1", type: "Integer" })
+        assert_includes(globals, { name: "$b", value: "2", type: "Integer" })
+
+        # Ruby defined globals
+        assert_includes(globals, { name: "$VERBOSE", value: "false", type: "FalseClass" })
+        assert_includes(globals, { name: "$stdout", value: "#<IO:<STDOUT>>", type: "IO" })
+
+        req_terminate_debuggee
+      end
+    end
+  end
+
+  class CDPVariablesTest < ProtocolTestCase
+    PROGRAM = <<~RUBY
+      1| $a = 1
+      2| $b = 2
+      3| $c = 3
+    RUBY
+
+    def test_eval_evaluates_global_variables
+      run_protocol_scenario PROGRAM, dap: false do
+        req_add_breakpoint 3
+        req_continue
+
+        globals = gather_variables(type: "global")
+
+        # User defined globals
+        assert_includes(globals, { name: "$a", value: "1", type: "Number" })
+        assert_includes(globals, { name: "$b", value: "2", type: "Number" })
+
+        # Ruby defined globals
+        assert_includes(globals, { name: "$VERBOSE", value: "false", type: "Boolean" })
+        assert_includes(globals, { name: "$stdout", value: "#<IO:<STDOUT>>", type: "Object" })
+
+        req_terminate_debuggee
+      end
+    end
+  end
+
+  class DAPInstanceVariableTest < ProtocolTestCase
+    PROGRAM = <<~RUBY
+      1| @a = 1
+      2| @c = 3
+      3| @b = 2
+      4| @z = 4
```

Comment:
Also a nitpick: I know it's because the breakpoint would stop the beginning of the line so `@z` wouldn't show up in the response. But at the first glance it's not very obvious. Maybe we can make it something else like `__LINE__` to avoid the confusion?